### PR TITLE
feat: avoid fencing instance unnecessarily in volume snapshots

### LINKED ISSUE #2853: Volume snapshot reconciler should fence instances only if there are no snapshots
The volumesnapshot reconciler starts by fencing the target pod right away.
This can lead to trouble if the top reconciliation loop has an error downstream that prevents a backup from being marked as completed.
In such case, the backup reconciler will get called once again, and will start by fencing the instance, then doing a listing where it might well detect that there are voluesnapshots already created, and would then proceed to unfence the instance.
We should only be fencing the pod if we have positively identified that there are no volumesnapshots created.

---

The logic of the snapshot backup reconciler is:

1. Step 1: fencing
2. Step 2: create snapshots
3. Step 3: wait for snapshots to be ready

as implemented in the volumesnapshot reconciler code (`pkg/reconciler/backup/volumesnapshot/resources.go`):

``` go
// Execute the volume snapshot of the given cluster instance
func (se *Reconciler) Execute(
```

### PR DESCRIPTION
**Pull request by jsilvela**
Closes #2853

### CODE REVIEW

File: pkg/reconciler/backup/volumesnapshot/reconciler.go [hunk 1]
```diff
@@ -123,10 +123,20 @@ func (se *Reconciler) Execute(
 	targetPod *corev1.Pod,
 	pvcs []corev1.PersistentVolumeClaim,
 ) (*ctrl.Result, error) {
+	if cluster.Spec.Backup == nil || cluster.Spec.Backup.VolumeSnapshot == nil {
```
Comment (YanniHu1996) [inline pkg/reconciler/backup/volumesnapshot/reconciler.go:126]: we don't need to check the `Backup` here, the previous code handles this already, 
please refer to https://github.com/cloudnative-pg/cloudnative-pg/blob/main/controllers/backup_controller.go#L194

Comment (jsilvela) [reply pkg/reconciler/backup/volumesnapshot/reconciler.go:126]: The problem is that the unit tests won't go through that previous checking.
If you take it out and you test the code with a cluster with no backup, you get a panic.

Comment (YanniHu1996) [reply pkg/reconciler/backup/volumesnapshot/reconciler.go:126]: thanks, @jsilvela , it makes sence

**Review by YanniHu1996 — APPROVED** [review]

**Review by armru — APPROVED** [review]

**Review by mnencia — APPROVED** [review]

**Comment by mnencia** [comment]
/ok-to-merge E2E green on EDB fork https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/6314358265/job/17144644959

**Pull request merged**